Content.Server/Zombies/ZombieSystem.cs

```diff
@@ -156,24 +159,102 @@ private void DoGroan(EntityUid uid, ActiveZombieComponent component)
         }
 
+        private void OnZombieInit(EntityUid ent, ZombieComponent component, ComponentStartup _)
+        {
+            // Brute
+            component.HealingDamageSpecifier.DamageDict["Brute"] = -10;
+            component.HealingDamageSpecifier.DamageDict["Pierce"] = -10;
+            component.HealingDamageSpecifier.DamageDict["Blunt"] = -10;
+
+            // Burn
+            component.HealingDamageSpecifier.DamageDict["Heat"] = -10;
+            component.HealingDamageSpecifier.DamageDict["Shock"] = -10;
+            component.HealingDamageSpecifier.DamageDict["Cold"] = -10;
+
+            // Zombies do not receive any Toxic or Airloss damage
+        }
+
+        /// <summary>
+        /// Heals zombie if it is in bad state and not dead
+        /// </summary>
+        private void TryHealZombie(EntityUid ent, ZombieComponent component)
+        {
+            if (!TryComp<MobStateComponent>(ent, out var state) || !TryComp<DamageableComponent>(ent, out var damage))
+            {
+                return;
+            }
+
+
+            if (state.CurrentState == DamageState.Critical)
+            {
+                _damageable.TryChangeDamage(ent, component.HealingDamageSpecifier, true);
+                return;
+            }
+
+            if (state.CurrentState != DamageState.Alive)
+                return;
+            var dmg = damage.TotalDamage;
+
+            // Always zero, I guess?
+            var aliveThreshold = 0f;
+            var critStateData = _state.GetEarliestCriticalState(state, dmg);
+            var nextThreshold = FixedPoint2.Zero;
+            if (critStateData != null)
+            {
+                nextThreshold = critStateData.Value.threshold;
+            }
+            else
+            {
+                var deadStateData = _state.GetEarliestDeadState(state, dmg);
+                if (deadStateData == null)
+                {
+                    return;
+                }
+
+                nextThreshold = deadStateData.Value.threshold;
+            }
+
+
+            // Healing to 20% of HP. Not too big and not too small.
+            var healingThreshold = aliveThreshold + (nextThreshold - aliveThreshold) * 0.8;
+
+            if (dmg > healingThreshold)
+            {
+                _damageable.TryChangeDamage(ent, component.HealingDamageSpecifier, true);
+            }
+        }
+
         public override void Update(float frameTime)
         {
             base.Update(frameTime);
 
-            foreach (var zombiecomp in EntityQuery<ActiveZombieComponent>())
+            foreach (var zombiecomp in EntityQuery<ZombieComponent>())
             {
-                zombiecomp.Accumulator += frameTime;
-                zombiecomp.LastDamageGroanCooldown -= frameTime;
+                // Only zombies that are alive have this thing
+                if (TryComp<ActiveZombieComponent>(zombiecomp.Owner, out var activeComp))
+                {
```

Comment:
You've replaced this with a slower iterator, just use the overload that takes 2 components.

Reply:
Zombies in crit get their active component removed so it won't work

Reply:
I found out zombies always have Zombie component, but if they're state is not Alive, active component gets removed. So overload with 2 types won't work properly if zombie is in crit.

Reply:
Update the code to not iterate every zombie